### PR TITLE
Clear values when UNSAT

### PR DESCRIPTION
**Pull request by IgnaceBleukx**
There was some inconcistency regarding what to do when the solver found UNSAT.
Z3 already cleared the values of all known variables, which I think made sense.

Extended this behaviour to all other solvers too and updated documentation

### CODE REVIEW

File: cpmpy/solvers/choco.py [hunk 1]
```diff
@@ -175,6 +175,9 @@ def solve(self, time_limit=None, **kwargs):
             # translate objective
             if self.has_objective():
                 self.objective_value_ = sol.get_int_val(self.solver_var(self.obj))
+        else: # clear values of variables
+            for cpm_var in self.user_vars:
+                cpm_var.clear()
```
Comment (Dimosts) [inline cpmpy/solvers/choco.py:180]: I guess we also have to put None to objective value??

Also, it is possible that vars have already value, and then the user changes something and uses solveAll(). So, in that case we also need to empty them? As I see in the solveAll() code, it normally has the value of the last solution found...

File: cpmpy/solvers/exact.py [hunk 2]
```diff
@@ -127,7 +127,7 @@ def _fillVars(self):
         if not self.xct_solver.hasSolution():
             self.objective_value_ = None
```
Comment (Dimosts) [inline cpmpy/solvers/exact.py:128]: yes, here we do empty the objective value, so we should do it also in all solvers

File: cpmpy/solvers/z3.py [hunk 3]
```diff
@@ -190,9 +190,9 @@ def solve(self, time_limit=None, assumptions=[], **kwargs):
                 obj = self.z3_solver.objectives()[0]
                 self.objective_value_ = sol.evaluate(obj).as_long()
 
-        else:
+        else:  # clear values of variables
             for cpm_var in self.user_vars:
-                cpm_var._value = None # XXX, maybe all solvers should do this...
```
Comment (Dimosts) [inline cpmpy/solvers/z3.py:195]: maybe yes, indeed :)

File: tests/test_solvers.py [hunk 4]
```diff
@@ -753,6 +753,21 @@ def test_incremental_objective(self):
...
+
+        for name, cls in cp.SolverLookup.base_solvers():
+            if cls.supported() is False: # solver not supported
+                continue
+            self.assertTrue(sat_model.solve(solver=name))
+            for v in (x,y,z):
+                self.assertIsNotNone(v.value())
+            self.assertFalse(unsat_model.solve(solver=name))
+            for v in (x,y,z):
+                self.assertIsNone(v.value())
```
Comment (Dimosts) [inline tests/test_solvers.py:771]: tests also with solveAll() and objective after they are done?

**Review by Dimosts — CHANGES_REQUESTED** [review]
Good, indeed needed! 

My comments mainly have to do that we have to be consistent also with objective value (we should clear that too) and also when we do solveAll() and the count is 0, we should again clear the values.

File: cpmpy/solvers/exact.py (hunk 2 above)
Comment (IgnaceBleukx) [reply cpmpy/solvers/exact.py:128]: It is actually already there for all solvers, right before the translation of the solution

**Review by tias — COMMENTED** [review]
Yes, but why pay the cost of an extra 'clear()' lookup and function call?

A few lines above each added for loop, we already directly access ._value; I think we better set v._value = None given that the code above already assumes you know it exists, and then it is more lightweight...

**Pull request merged**